src/common/display/src/tree.rs

```diff
@@ -1,131 +1,49 @@
-            [] => Ok(()),
-            // One child.
-            [child] => {
-                // Child tree.
-                child.fmt_tree_indent_style(indent + 1, s)
-            }
-            // Two children.
-            [left, right] => {
-                left.fmt_tree_indent_style(indent + 1, s)?;
-                right.fmt_tree_indent_style(indent + 1, s)
-            }
-            _ => unreachable!("Max two child nodes expected, got {}", children.len()),
-        }
+    /// Get the human-readable name of this node.
+    fn get_name(&self) -> String {
+        std::any::type_name::<Self>().to_string()
     }
 
-    // Print the tree recursively, and illustrate the tree structure in the same style as `git log --graph`.
-    // `depth` is the number of forks in this node's ancestors.
-    fn fmt_tree_gitstyle(&self, depth: usize, s: &mut String, simple: bool) -> fmt::Result {
-        // Print the current node.
-        // e.g. | | * <node contents line 1>
-        //      | | | <node contents line 2>
-        let lines = if simple {
-            vec![self.get_name()]
-        } else {
-            self.get_multiline_representation()
-        };
-        use terminal_size::{terminal_size, Width};
-        let size = terminal_size();
-        let term_width = if let Some((Width(w), _)) = size {
-            w as usize
-        } else {
-            88usize
-        };
-
-        let mut counter = 0;
-        for val in lines.iter() {
-            let base_characters = depth * 2;
-            let expected_chars = (term_width - base_characters - 8).max(8);
-            let sublines = textwrap::wrap(val, expected_chars);
-
-            for (i, sb) in sublines.iter().enumerate() {
-                self.fmt_depth(depth, s)?;
-                match counter {
-                    0 => write!(s, "* ")?,
-                    _ => write!(s, "|   ")?,
-                }
-                counter += 1;
-                match i {
-                    0 => writeln!(s, "{sb}")?,
-                    _ => writeln!(s, "  {sb}")?,
-                }
-            }
-        }
-
-        // Recursively handle children.
-        let children = self.get_children();
-        match &children[..] {
-            // No children - stop printing.
-            [] => Ok(()),
-            // One child - print leg, then print the child tree.
-            [child] => {
-                // Leg: e.g. | | |
-                self.fmt_depth(depth, s)?;
-                writeln!(s, "|")?;
-
-                // Child tree.
-                child.fmt_tree_gitstyle(depth, s, simple)
-            }
-            // Two children - print legs, print right child indented, print left child.
-            [left, right] => {
-                // Legs: e.g. | | |\
-                self.fmt_depth(depth, s)?;
-                writeln!(s, "|\\")?;
+    /// Required method: Get the children of the self node.
+    fn get_children(&self) -> Vec<Arc<dyn TreeDisplay>>;
```

Comment:
The key issues with TreeDisplay is that it requires `fn get_children(&self) -> Vec<Arc<dyn TreeDisplay>>;` to be implemented. However our representation for the Pipeline is storing` Box<dyn Trait>` so it makes it a bit hard. Do you think it would be possible to adapt this to `fn get_children(&self) -> Vec<&dyn TreeDisplay>;`

Reply:
yeah I think that should be pretty do-able

Reply:
@samster25 I just made these changes, LMK if there's anything else!